nix/shell.go

```diff
@@ -0,0 +1,181 @@
+	case "zsh":
+		sh.name = shZsh
+		sh.userShellrcPath = rcfilePath(".zshrc")
+	case "ksh":
+		sh.name = shKsh
+		sh.userShellrcPath = rcfilePath(".kshrc")
+	case "dash", "ash", "sh":
+		sh.name = shPosix
+		sh.userShellrcPath = os.Getenv("ENV")
+
+		// Just make up a name if there isn't already an init file set
+		// so we have somewhere to put a new one.
+		if sh.userShellrcPath == "" {
+			sh.userShellrcPath = ".shinit"
+		}
+	default:
+		sh.name = shUnknown
+	}
+	debug.Log("Detected shell: %s", sh.binPath)
+	debug.Log("Recognized shell as: %s", sh.binPath)
+	debug.Log("Looking for user's shell init file at: %s", sh.userShellrcPath)
+	return sh, nil
+}
+
+// rcfilePath returns the absolute path for an rcfile, which is usually in the
+// user's home directory. It doesn't guarantee that the file exists.
+func rcfilePath(basename string) string {
+	home, err := os.UserHomeDir()
+	if err != nil {
+		return ""
+	}
+	return filepath.Join(home, basename)
+}
+
+func (s *Shell) Run(nixPath string) error {
+	// Launch a fallback shell if we couldn't find the path to the user's
+	// default shell.
+	if s.binPath == "" {
+		cmd := exec.Command("nix-shell", nixPath)
+		cmd.Stdin = os.Stdin
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+
+		debug.Log("Unrecognized user shell, falling back to: %v", cmd.Args)
+		return errors.WithStack(cmd.Run())
+	}
+
+	cmd := exec.Command("nix-shell", nixPath)
+	cmd.Args = append(cmd.Args, "--pure", "--command", s.execCommand())
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	debug.Log("Executing nix-shell command: %v", cmd.Args)
+	return errors.WithStack(cmd.Run())
+}
+
+// execCommand is a command that replaces the current shell with s.
+func (s *Shell) execCommand() string {
+	shellrc, err := writeDevboxShellrc(s.userShellrcPath, s.UserInitHook)
+	if err != nil {
+		debug.Log("Failed to write devbox shellrc: %v", err)
+		return "exec " + s.binPath
+	}
+
+	switch s.name {
+	case shBash:
+		return fmt.Sprintf(`exec /usr/bin/env ORIGINAL_PATH="%s" %s --rcfile "%s"`,
+			os.Getenv("PATH"), s.binPath, shellrc)
+	case shZsh:
+		return fmt.Sprintf(`exec /usr/bin/env ORIGINAL_PATH="%s" ZDOTDIR="%s" %s`,
+			os.Getenv("PATH"), filepath.Dir(shellrc), s.binPath)
+	case shKsh, shPosix:
+		return fmt.Sprintf(`exec /usr/bin/env ORIGINAL_PATH="%s" ENV="%s" %s `,
+			os.Getenv("PATH"), shellrc, s.binPath)
+	default:
+		return "exec " + s.binPath
+	}
+}
+
+func writeDevboxShellrc(userShellrcPath string, userHook string) (path string, err error) {
+	// We need a temp dir (as opposed to a temp file) because zsh uses
+	// ZDOTDIR to point to a new directory containing the .zshrc.
+	tmp, err := os.MkdirTemp("", "devbox")
+	if err != nil {
+		return "", fmt.Errorf("create temp dir for shell init file: %v", err)
+	}
+
+	// This is a best-effort to include the user's existing shellrc. If we
+	// can't read it, then just omit it from the devbox shellrc.
+	userShellrc, _ := os.ReadFile(userShellrcPath)
```

Comment:
maybe only ignore error for `os.IsNotExist(err)`?
and set userShellrc to be empty-string in that case?

Reply:
I think you're right here. It might be possible to return an error and a partially read file. I'll update it to explicitly set `userShellrc` to empty if there's an error so we don't risk running part of the user's shellrc.